File: src/services/moduleService.js
import {
  collection,
  doc,
  getDoc,
  getDocs,
  addDoc,
  updateDoc,
  deleteDoc,
  query,
  where,
  orderBy,
  serverTimestamp,
} from "firebase/firestore";
import { db } from "../firebase";

const moduleService = {
  // Get all modules
  async getAllModules() {
    try {
      const modulesRef = collection(db, "modules");
      const querySnapshot = await getDocs(modulesRef);
      const modules = querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));
      return modules.sort((a, b) => (a.order || 0) - (b.order || 0));
    } catch (error) {
      console.error("Error getting all modules:", error);
      throw error;
    }
  },

  // Get all modules for a course
  async getModulesByCourseId(courseId) {
    try {
      const modulesRef = collection(db, "modules");
      const q = query(modulesRef, where("courseId", "==", courseId));
      const querySnapshot = await getDocs(q);
      const modules = querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));
      // Sort in memory
      return modules.sort((a, b) => (a.order || 0) - (b.order || 0));
    } catch (error) {
      console.error("Error getting modules:", error);
      throw error;
    }
  },

  // Get lessons for a specific module
  async getLessonsByModule(courseId, moduleId) {
    try {
      const lessonsRef = collection(db, "lessons");
      const q = query(
        lessonsRef,
        where("courseId", "==", courseId),
        where("moduleId", "==", moduleId)
      );
      const querySnapshot = await getDocs(q);
      const lessons = querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));
      // Sort in memory
      return lessons.sort((a, b) => (a.order || 0) - (b.order || 0));
    } catch (error) {
      console.error("Error getting lessons:", error);
      throw error;
    }
  },

  // Get a single module by ID
  async getModuleById(moduleId) {
    try {
      const moduleRef = doc(db, "modules", moduleId);
      const moduleDoc = await getDoc(moduleRef);
      if (!moduleDoc.exists()) {
        throw new Error("Module not found");
      }
      return {
        id: moduleDoc.id,
        ...moduleDoc.data(),
      };
    } catch (error) {
      console.error("Error getting module:", error);
      throw error;
    }
  },

  // Create a new module
  async createModule(courseId, moduleData) {
    try {
      const modulesRef = collection(db, "modules");
      const newModule = {
        ...moduleData,
        courseId,
        createdAt: serverTimestamp(),
        updatedAt: serverTimestamp(),
      };
      const docRef = await addDoc(modulesRef, newModule);
      return {
        id: docRef.id,
        ...newModule,
      };
    } catch (error) {
      console.error("Error creating module:", error);
      throw error;
    }
  },

  // Update a module
  async updateModule(moduleId, moduleData) {
    try {
      const moduleRef = doc(db, "modules", moduleId);
      const updateData = {
        ...moduleData,
        updatedAt: serverTimestamp(),
      };
      await updateDoc(moduleRef, updateData);
      return {
        id: moduleId,
        ...updateData,
      };
    } catch (error) {
      console.error("Error updating module:", error);
      throw error;
    }
  },

  // Delete a module
  async deleteModule(moduleId) {
    try {
      const moduleRef = doc(db, "modules", moduleId);
      await deleteDoc(moduleRef);
      return true;
    } catch (error) {
      console.error("Error deleting module:", error);
      throw error;
    }
  },

  // Reorder modules
  async reorderModules(moduleIds) {
    try {
      const batch = db.batch();
      moduleIds.forEach((moduleId, index) => {
        const moduleRef = doc(db, "modules", moduleId);
        batch.update(moduleRef, {
          order: index,
          updatedAt: serverTimestamp(),
        });
      });
      await batch.commit();
      return true;
    } catch (error) {
      console.error("Error reordering modules:", error);
      throw error;
    }
  },
};

export default moduleService;

File: src/student-ui/to react/lib/services/module_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import '../../../../../migrate/lib/models/module_model.dart';
import '../../../../../migrate/lib/models/lesson_model.dart';

class ModuleService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final String _modulesCollection = 'modules';

  // Get all modules for a course
  Future<List<Module>> getModulesByCourse(String courseId) async {
    try {
      debugPrint('=== Fetching Modules ===');
      debugPrint('Course ID: $courseId');

      final querySnapshot = await _firestore
          .collection(_modulesCollection)
          .where('courseId', isEqualTo: courseId)
          .get();

      debugPrint(
        'Query completed. Documents found: ${querySnapshot.docs.length}',
      );

      if (querySnapshot.docs.isEmpty) {
        debugPrint('No modules found for course: $courseId');
        return [];
      }

      final modules = querySnapshot.docs.map((doc) {
        debugPrint('Processing module document: ${doc.id}');
        debugPrint('Module data: ${doc.data()}');
        return Module.fromFirestore(doc);
      }).toList();

      // Sort modules by order after fetching
      modules.sort((a, b) => a.order.compareTo(b.order));

      debugPrint('Successfully processed ${modules.length} modules');
      return modules;
    } catch (e, stackTrace) {
      debugPrint('Error getting modules: $e');
      debugPrint('Stack trace: $stackTrace');
      rethrow;
    }
  }

  // Get a single module
  Future<Module?> getModule(String moduleId) async {
    try {
      final doc = await _firestore
          .collection(_modulesCollection)
          .doc(moduleId)
          .get();
      if (!doc.exists) return null;
      return Module.fromFirestore(doc);
    } catch (e) {
      debugPrint('Error getting module: $e');
      rethrow;
    }
  }

  // Create a new module
  Future<Module> createModule(String courseId, Module module) async {
    try {
      final docRef = await _firestore
          .collection(_modulesCollection)
          .add(module.toMap());
      final doc = await docRef.get();
      return Module.fromFirestore(doc);
    } catch (e) {
      debugPrint('Error creating module: $e');
      rethrow;
    }
  }

  // Update an existing module
  Future<Module> updateModule(String moduleId, Module module) async {
    try {
      final docRef = _firestore.collection(_modulesCollection).doc(moduleId);
      await docRef.update(module.toMap());
      final doc = await docRef.get();
      return Module.fromFirestore(doc);
    } catch (e) {
      debugPrint('Error updating module: $e');
      rethrow;
    }
  }

  // Delete a module
  Future<void> deleteModule(String moduleId) async {
    try {
      await _firestore.collection(_modulesCollection).doc(moduleId).delete();
    } catch (e) {
      debugPrint('Error deleting module: $e');
      rethrow;
    }
  }

  // Get modules stream for real-time updates
  Stream<List<Module>> getModulesStream(String courseId) {
    return _firestore
        .collection(_modulesCollection)
        .where('courseId', isEqualTo: courseId)
        .orderBy('order')
        .snapshots()
        .map(
          (snapshot) =>
              snapshot.docs.map((doc) => Module.fromFirestore(doc)).toList(),
        );
  }

  // Get all lessons for a module
  Stream<List<LessonModel>> getLessons(String courseId, String moduleId) {
    return _firestore
        .collection('courses')
        .doc(courseId)
        .collection('modules')
        .doc(moduleId)
        .collection('lessons')
        .orderBy('order')
        .snapshots()
        .map((snapshot) {
          return snapshot.docs
              .map((doc) => LessonModel.fromFirestore(doc))
              .toList();
        });
  }

  // Get a single lesson
  Future<LessonModel?> getLesson(
    String courseId,
    String moduleId,
    String lessonId,
  ) async {
    final doc = await _firestore
        .collection('courses')
        .doc(courseId)
        .collection('modules')
        .doc(moduleId)
        .collection('lessons')
        .doc(lessonId)
        .get();

    if (!doc.exists) return null;
    return LessonModel.fromFirestore(doc);
  }
}
